File: Sudoku/Solve/Solver.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;

namespace Sudoku
{
    /// <summary>
    /// Don't judge my by this ancient code. I deliberately haven't tidied it up.
    /// </summary>
    public class Solver
    {
        int[, ] SudokuGrid = new int[9, 9];
        int[, , ] Possibilities = new int[9, 9, 9];
        bool Restart;
        bool Exit = false;
        bool Change;
        bool Contradiction;

        public int[,] Solve(int?[,] Boxes)
        {
            Restart = false;
            int i, j;
            for (i = 0; i < 9; i++)
            {
                for (j = 0; j < 9; j++)
                {
                    if (Boxes[i, j] == null)
                    {
                        SudokuGrid[i, j] = 0;
                    }
                    else
                    {
                        SudokuGrid[i, j] = Boxes[i, j].Value;
                    }
                }
            }
            //if (Restart != true)
            //{
                return FindSolution();
            //}
        }

        private int[,] FindSolution()
        {
            int i, j, k;
            if (!IsValid())
            {
                throw new InvalidSudokuException();
            }
            for (i = 0; i < 9; i++)          /* anything is possible                  */
                for (j = 0; j < 9; j++)
                    for (k = 0; k < 9; k++)
                    {
                        Possibilities[i, j, k] = 1;
                    }
            while (!Finished())          /* the actual program: alternate between */
            {         
                Basic();                     /* basic techniques...  */
                if (Exit)
                    break;
                if (!Change)
                    if (!MakeGuess())  /* ...and guessing                       */
                    {
                        try
                        {
                            Guess2();
                        }
                        catch (Exception)
                        {
                            throw new UnableToSolveException();
                        }
                        break;
                    }
            }
            return SudokuGrid;
        }

        private bool IsValid()  /* checks the current grid is valid */
        {     
            int i,j;
            int[] list1 = new int[10];
            int[] list2 = new int[10];
            int[] list3 = new int[10];
            for(i=0 ; i<9 ; i++) 
            {     /* i is a row, columnm or box */
                for (j = 0; j < 10; j++)
                {
                    list1[j] = 0;
                    list2[j] = 0;
                    list3[j] = 0;            /* lists mark off each number as it's found */
                }
                for(j=0 ; j<9 ; j++) 
                {   /* j are the 9 elements of i */
                    list1[SudokuGrid[i, j]]++;                           /* rows */
                    if(list1[SudokuGrid[i, j]]>1 && SudokuGrid[i, j]>0)
	                    return false;
                    list2[SudokuGrid[j, i]]++;                           /* columns */
                    if(list2[SudokuGrid[j, i]]>1 && SudokuGrid[j, i]>0)
	                    return false;
                    list3[ SudokuGrid[(i/3)*3 + j%3, (i*3)%9 + j/3] ]++; /* boxes */
                    if(list3[ SudokuGrid[(i/3)*3 + j%3, (i*3)%9 + j/3] ] > 1 
	                    && SudokuGrid[(i/3)*3 + j%3, (i*3)%9 + j/3] > 0)
	                    return false;
                }
            }
        return true;
        }

        private void UpdateGrid() /* puts in a number if it is the only possibility */
        {  
            int i, j, k;
            int sum;
            for (i = 0; i < 9; i++)
            {
                for (j = 0; j < 9; j++)
                    if (SudokuGrid[i, j] == 0)
                    {
                        sum = 0;
                        for (k = 0; k < 9; k++)
                            sum += Possibilities[i, j, k];
                        if (sum == 1)
                            for (k = 0; k < 9; k++)
                                if (Possibilities[i, j, k] == 1)
                                {
                                    SudokuGrid[i, j] = k + 1;
                                    Change = true;
                                    break;
                                }
                        if (sum < 1)
                        {
                            //printf("\nUpdate error: no possibilities found for [%d, %d]\n", i, j);
                            Contradiction = true;
                            return;
                        }
                    }
            }
        } /* end of update */

        private void UpdateGrid2() /* adds numbers if there is only one place they can go */
        {
            int n, i, j, k;
            int[] sum = new int[3];
            for (n = 1; n < 10; n++)         /* n a number in the grid */
            {
                for (i = 0; i < 9; i++)
                {      /* i a column, row, or box */
                    for (k = 0; k < 3; k++)      /* 0 column, 1 row, 2 box */
                        sum[k] = 0;
                    for (j = 0; j < 9; j++)
                    {    /* j an element of i */
                        sum[0] += Possibilities[i, j, n - 1];
                        sum[1] += Possibilities[j, i, n - 1];
                        sum[2] += Possibilities[(i / 3) * 3 + j % 3, (i * 3) % 9 + j / 3, n - 1];
                    }
                    if (sum[0] == 1)
                        for (j = 0; j < 9; j++)
                            if (Possibilities[i, j, n - 1] == 1)
                            {
                                if (SudokuGrid[i, j] == 0)
                                    Change = true;
                                SudokuGrid[i, j] = n;
                                break;
                            }
                    if (sum[0] < 1)
                    {
                        //printf("\nUpdate2 error: nowhere to put %d in column %d\n", n, i);
                        Contradiction = true;
                        return;
                    }
                    if (sum[1] == 1)
                        for (j = 0; j < 9; j++)
                            if (Possibilities[j, i, n - 1] == 1 && SudokuGrid[j, i] == 0)
                            {
                                SudokuGrid[j, i] = n;
                                Change = true;
                                break;
                            }
                    if (sum[1] < 1)
                    {
                        //printf("\nUpdate2 error: nowhere to put %d in row %d\n", n, i);
                        Contradiction = true;
                        return;
                    }
                    if (sum[2] == 1)
                        for (j = 0; j < 9; j++)
                            if (Possibilities[(i / 3) * 3 + j % 3, (i * 3) % 9 + j / 3, n - 1] == 1)
                            {
                                if (SudokuGrid[(i / 3) * 3 + j % 3, (i * 3) % 9 + j / 3] == 0)
                                    Change = true;
                                SudokuGrid[(i / 3) * 3 + j % 3, (i * 3) % 9 + j / 3] = n;
                                break;
                            }
                    if (sum[2] < 1)
                    {
                        //printf("\nUpdate2 error: nowhere to put %d in box %d\n", n, i);
                        Contradiction = true;
                        return;
                    }
                }
            }
        }

        private void Reduce() 
        {       /* reduces possibilities according to a[, ] */
            int i, j, k, l;
            for (i = 0; i < 9; i++)
                for (j = 0; j < 9; j++)
                    if (SudokuGrid[i, j] > 0)
                    {
                        for (k = 0; k < 9; k++)
                        {
                            Possibilities[i, k, SudokuGrid[i, j] - 1] = 0;
                            Possibilities[k, j, SudokuGrid[i, j] - 1] = 0;
                            Possibilities[i, j, k] = 0;
                        }
                        for (k = 0; k < 3; k++)
                            for (l = 0; l < 3; l++)
                                /* box coordinate i/3,j/3; top left of box is (i/3)*3,(j/3)*3 */
                                Possibilities[(i / 3) * 3 + k, (j / 3) * 3 + l, SudokuGrid[i, j] - 1] = 0;
                        Possibilities[i, j, SudokuGrid[i, j] - 1] = 1;
                    }
        }

        private bool Finished() /* checks if the sudoku is finished yet */
        {    
            int i, j;
            for (i = 0; i < 9; i++)
                for (j = 0; j < 9; j++)
                    if (SudokuGrid[i, j] == 0)
                        return false;
            //label2.Text = "Solved it!";
            return true;
        }

        private void Basic() /* applies the basic methods repeatedly */
        {             
            int pass=0;                 /* number of passes */
            Contradiction=false;
            Change=true;
            while(Change && !Finished()) 
            {
                Reduce();
                Change=false;
                UpdateGrid();
                UpdateGrid2();
                if(!IsValid() || Contradiction) 
                {
                    //Display();
                    //outposs();
                    //label2.Text = "I screwed it up!";
                    Exit = true;
                    return;
                }
                pass++;
            } /* end while loop */
            //printf("\nAfter %d basic passes we have:",pass);
            // ??? Display();
            if (!Change)
            {
                //label2.Text = "Basic passes exhausted.";
            }
        }

        private bool MakeGuess()  /* when other methods fail, makes a conjecture */
        {  
            int n,i,j,k;         /* and seeks a contradiction/solution          */
            int row,col;
            bool end=false;
            int pass=1,maxpass=100;
            int[,] StoreGrid = new int[9,9];
            int[,,] StorePoss = new int[9, 9, 9];
            //label2.Text = "\nBeginning makeguess function.";
            for (row = 0; row < 9; row++)
            {
                for (col = 0; col < 9; col++)
                {
                    if (SudokuGrid[row, col] == 0)        /* conjecture that this cell... */
                    {
                        for (n = 1; n <= 9; n++)   /* ...contains n                */
                        {
                            if (Possibilities[row, col, n - 1] == 1)
                            {
                                //printf("\nConjecturing that [%d, %d] contains %d...",row,col,n);
                                for (i = 0; i < 9; i++) /* fill in the storage arrays */
                                    for (j = 0; j < 9; j++)
                                    {
                                        StoreGrid[i, j] = SudokuGrid[i, j];
                                        for (k = 0; k < 9; k++)
                                            StorePoss[i, j, k] = Possibilities[i, j, k];
                                    }
                                SudokuGrid[row, col] = n; /* make the conjecture! */
                                Contradiction = false;
                                while (pass <= maxpass)
                                {
                                    Reduce();
                                    Change = false;
                                    UpdateGrid();
                                    if (!Contradiction)
                                        UpdateGrid2();
                                    if (Finished())
                                    {
                                        //printf("done, leading to the solution:");
                                        // FINISHED Display();
                                        return true;
                                    }
                                    if (!IsValid() || Contradiction)
                                    {
                                        StorePoss[row, col, n - 1] = 0;
                                        end = true;
                                        //printf("done, ruled it out.");
                                        break;
                                    }
                                    if (!Change)
                                    {
                                        //printf("done, got stuck.");
                                        break;
                                    }
                                    pass++;
                                } /* end while loop */
                                for (i = 0; i < 9; i++) /* restore a[, ] and poss[, , ] */
                                    for (j = 0; j < 9; j++)
                                    {
                                        SudokuGrid[i, j] = StoreGrid[i, j];
                                        for (k = 0; k < 9; k++)
                                            Possibilities[i, j, k] = StorePoss[i, j, k];
                                    }
                            } /* end of current conjecture "if n is a poss..."; end n loop */
                        }
                        if (end)
                        {
                            return true;
                        }
                    }
                } /* end col loop */
            }
            return end;
        } /* end of makeguess */

        private void Guess2() /* desperate now - try the long way */
        {
            int i, j, k, n;
            int row, col;
            int[,] Store = new int[9, 9];
            int[, ,] StorePoss = new int[9, 9, 9];
            //label2.Text = "Beginning guess2 function.";
            for (i = 0; i < 9; i++) /* fill in the storage arrays */
            {
                for (j = 0; j < 9; j++)
                {
                    Store[i, j] = SudokuGrid[i, j];
                    for (k = 0; k < 9; k++)
                    {
                        StorePoss[i, j, k] = Possibilities[i, j, k];
                    }
                }
            }
            row = 0;
            col = 0;
            n = 1;
            while (row < 9)
            {
                while(col<9)
                {
                    while(n<10)
                    {
                        if (Possibilities[row, col, n-1] == 1)
                        {
                            SudokuGrid[row, col] = n;
                            n = 1;
                            break;
                        }
                        n++;
                    }
                    if (n >= 10) //Occurs only if we didn't find anything
                    {
                        //Backtrack here
                        if (col == 0)
                        {
                            row--;
                            col = 8;
                        }
                        else
                        {
                            col--;
                        }
                        n = SudokuGrid[row, col] + 1;
                        SudokuGrid[row, col] = 0;
                        for (i = 0; i < 9; ++i)
                        {
                            for (j = 0; j < 9; ++j)
                            {
                                for (k = 0; k < 9; ++k)
                                {
                                    Possibilities[i, j, k] = StorePoss[i, j, k];
                                }
                            }
                        }
                        Reduce();
                    }
                    else
                    {
                        Reduce();
                        //Basic();
                        //Reduce();
                        col++;
                        n = 1;
                    }
                }
                row++;
                col = 0;
            }
            // FINISHED Display();
            //label2.Text = "Solved it!";
        }

    }
}

File: src/app/logic/solver.ts
import { GridValue } from "../types";

const initialiseArray = (length: number, width: number, depth?: number) => {
    const ret = new Array(length);
    for (let i = 0; i < length; ++i) {
        ret[i] = new Array(width);
        if (depth) {
            for (let j = 0; j < width; ++j) {
                ret[i][j] = new Array(depth);
            }
        }
    }
    return ret;
};

export default class Solver {
    private sudokuGrid: number[][] = initialiseArray(9, 9);
    private possibilities: number[][][]  = initialiseArray(9, 9, 9);
    private exit: boolean = false;
    private change: boolean = false;
    private contradiction: boolean = false;
    
    private log: (message: string) => void;

    public constructor(log?: (message: string) => void) {
        this.log = log || console.log;
    }

    public solve(input: GridValue): GridValue {
        for (let i = 0; i < 9; i++)
        {
            for (let j = 0; j < 9; j++)
            {
                this.sudokuGrid[i][j] = input[i][j] || 0;
            }
        }
        return this.findSolution(); // shouldn't have to convert zeros to nulls
    }

    private findSolution()
    {
        if (!this.isValid(this.sudokuGrid))
        {
            throw new Error("Invalid Sudoku");
        }
        for (let i = 0; i < 9; i++)          /* anything is possible                  */
            for (let j = 0; j < 9; j++)
                for (let k = 0; k < 9; k++)
                {
                    this.possibilities[i][j][k] = 1;
                }
        while (!this.finished())        /* the actual program: alternate between */
        {         
            this.basic();               /* basic techniques...  */
            if (this.exit)
                break;
            if (!this.change)
                if (!this.makeGuess())  /* ...and guessing                       */
                {
                    try
                    {
                        this.guess2();
                    }
                    catch (Exception)
                    {
                        throw new Error("Failed to solve this one!");
                    }
                    break;
                }
        }
        return this.sudokuGrid;
    }

    public isValid(input: number[][])  /* checks the current grid is valid */
    {     
        let list1:number[] = new Array(10);
        let list2:number[] = new Array(10);
        let list3:number[] = new Array(10);
        for(let i=0 ; i<9 ; i++) 
        {     /* i is a row, columnm or box */
            for (let j = 0; j < 10; j++)
            {
                list1[j] = 0;
                list2[j] = 0;
                list3[j] = 0;            /* lists mark off each number as it's found */
            }
            for(let j=0 ; j<9 ; j++) 
            {   /* j are the 9 elements of i */
                list1[input[i][j]]++;                           /* rows */
                if(list1[input[i][j]]>1 && input[i][j]>0)
                    return false;
                list2[input[j][i]]++;                           /* columns */
                if(list2[input[j][i]]>1 && input[j][i]>0)
                    return false;
                list3[ input[Math.floor(i/3)*3 + j%3][(i*3)%9 + Math.floor(j/3)] ]++; /* boxes */
                if(list3[ input[Math.floor(i/3)*3 + j%3][(i*3)%9 + Math.floor(j/3)] ] > 1 
                    && input[Math.floor(i/3)*3 + j%3][(i*3)%9 + Math.floor(j/3)] > 0)
                    return false;
            }
        }
        return true;
    }

    private updateGrid() /* puts in a number if it is the only possibility */
    {  
        for (let i = 0; i < 9; i++)
        {
            for (let j = 0; j < 9; j++)
                if (this.sudokuGrid[i][j] == 0)
                {
                    let sum = 0;
                    for (let k = 0; k < 9; k++)
                        sum += this.possibilities[i][j][k];
                    if (sum == 1)
                        for (let k = 0; k < 9; k++)
                            if (this.possibilities[i][j][k] == 1)
                            {
                                this.sudokuGrid[i][j] = k + 1;
                                this.change = true;
                                break;
                            }
                    if (sum < 1)
                    {
                        this.log(`Update error: no possibilities found for [${i}, ${j}]`);
                        this.contradiction = true;
                        return;
                    }
                }
        }
    } /* end of update */

    private updateGrid2() /* adds numbers if there is only one place they can go */
    {
        const sum: number[] = new Array(3);
        for (let n = 1; n < 10; n++)         /* n a number in the grid */
        {
            for (let i = 0; i < 9; i++)
            {      /* i a column, row, or box */
                for (let k = 0; k < 3; k++)      /* 0 column, 1 row, 2 box */
                    sum[k] = 0;
                for (let j = 0; j < 9; j++)
                {    /* j an element of i */
                    sum[0] += this.possibilities[i][j][n - 1];
                    sum[1] += this.possibilities[j][i][n - 1];
                    sum[2] += this.possibilities[Math.floor(i / 3) * 3 + j % 3][(i * 3) % 9 + Math.floor(j/3)][n - 1];
                }
                if (sum[0] == 1)
                    for (let j = 0; j < 9; j++)
                        if (this.possibilities[i][j][n - 1] == 1)
                        {
                            if (this.sudokuGrid[i][j] == 0)
                                this.change = true;
                            this.sudokuGrid[i][j] = n;
                            break;
                        }
                if (sum[0] < 1)
                {
                    this.log(`Update2 error: nowhere to put ${n} in column ${i}`);
                    this.contradiction = true;
                    return;
                }
                if (sum[1] == 1)
                    for (let j = 0; j < 9; j++)
                        if (this.possibilities[j][i][n - 1] == 1 && this.sudokuGrid[j][i] == 0)
                        {
                            this.sudokuGrid[j][i] = n;
                            this.change = true;
                            break;
                        }
                if (sum[1] < 1)
                {
                    this.log(`Update2 error: nowhere to put ${n} in row ${i}`);
                    this.contradiction = true;
                    return;
                }
                if (sum[2] == 1)
                    for (let j = 0; j < 9; j++)
                        if (this.possibilities[Math.floor(i / 3) * 3 + j % 3][(i * 3) % 9 + Math.floor(j/3)][n - 1] == 1)
                        {
                            if (this.sudokuGrid[Math.floor(i / 3) * 3 + j % 3][(i * 3) % 9 + Math.floor(j/3)] == 0)
                                this.change = true;
                            this.sudokuGrid[Math.floor(i / 3) * 3 + j % 3][(i * 3) % 9 + Math.floor(j/3)] = n;
                            break;
                        }
                if (sum[2] < 1)
                {
                    this.log(`Update2 error: nowhere to put ${n} in box ${i}`);
                    this.contradiction = true;
                    return;
                }
            }
        }
    }

    private reduce() 
    {       /* reduces possibilities according to a[, ] */
        for (let i = 0; i < 9; i++)
            for (let j = 0; j < 9; j++)
                if (this.sudokuGrid[i][j] > 0)
                {
                    for (let k = 0; k < 9; k++)
                    {
                        this.possibilities[i][k][this.sudokuGrid[i][j] - 1] = 0;
                        this.possibilities[k][j][this.sudokuGrid[i][j] - 1] = 0;
                        this.possibilities[i][j][k] = 0;
                    }
                    for (let k = 0; k < 3; k++)
                        for (let l = 0; l < 3; l++)
                            /* box coordinate i/3,j/3; top left of box is (i/3)*3,(j/3)*3 */
                            this.possibilities[Math.floor(i / 3) * 3 + k][Math.floor(j/3) * 3 + l][this.sudokuGrid[i][j] - 1] = 0;
                    this.possibilities[i][j][this.sudokuGrid[i][j] - 1] = 1;
                }
    }

    private finished() /* checks if the sudoku is finished yet */
    {    
        for (let i = 0; i < 9; i++)
            for (let j = 0; j < 9; j++)
                if (this.sudokuGrid[i][j] == 0)
                    return false;
        this.log("Solved it! (Whole grid is complete)");
        return true;
    }

    private basic() /* applies the basic methods repeatedly */
    {             
        let pass = 0;                 /* number of passes */
        this.contradiction=false;
        this.change=true;
        while(this.change && !this.finished()) 
        {
            this.reduce();
            this.change = false;
            this.updateGrid();
            this.updateGrid2();
            if(!this.isValid(this.sudokuGrid) || this.contradiction) 
            {
                // We might have screwed it up, but assume there is an inherent contradiction
                throw new Error("No solution exists");
            }
            ++pass;
        } /* end while loop */
        
        if (!this.change)
        {
            this.log(`Basic passes exhausted after ${pass} passes.`);
        }
    }

    private makeGuess() /* when other methods fail, makes a conjecture */
    {  
        let end = false;
        let pass = 1, maxpass = 100;
        let storeGrid = initialiseArray(9, 9);
        let storePoss = initialiseArray(9, 9, 9);
        //label2.Text = "\nBeginning this.makeGuess() function.";
        for (let row = 0; row < 9; row++)
        {
            for (let col = 0; col < 9; col++)
            {
                if (this.sudokuGrid[row][col] == 0)        /* conjecture that this cell... */
                {
                    for (let n = 1; n <= 9; n++)   /* ...contains n                */
                    {
                        if (this.possibilities[row][col][n - 1] == 1)
                        {
                            this.log(`Conjecturing that [${row}, ${col}] contains ${n}...`);
                            for (let i = 0; i < 9; i++) /* fill in the storage arrays */
                                for (let j = 0; j < 9; j++)
                                {
                                    storeGrid[i][j] = this.sudokuGrid[i][j];
                                    for (let k = 0; k < 9; k++)
                                        storePoss[i][j][k] = this.possibilities[i][j][k];
                                }
                            this.sudokuGrid[row][col] = n; /* make the conjecture! */
                            this.contradiction = false;
                            while (pass <= maxpass)
                            {
                                this.reduce();
                                this.change = false;
                                this.updateGrid();
                                if (!this.contradiction)
                                    this.updateGrid2();
                                if (this.finished())
                                {
                                    this.log("done, leading to the solution");
                                    // FINISHED Display();
                                    return true;
                                }
                                if (!this.isValid(this.sudokuGrid) || this.contradiction)
                                {
                                    storePoss[row][col][n - 1] = 0;
                                    end = true;
                                    this.log("done, ruled it out.");
                                    break;
                                }
                                if (!this.change)
                                {
                                    this.log("done, got stuck.");
                                    break;
                                }
                                pass++;
                            } /* end while loop */
                            for (let i = 0; i < 9; i++) /* restore a[, ] and poss[, , ] */
                                for (let j = 0; j < 9; j++)
                                {
                                    this.sudokuGrid[i][j] = storeGrid[i][j];
                                    for (let k = 0; k < 9; k++)
                                        this.possibilities[i][j][k] = storePoss[i][j][k];
                                }
                        } /* end of current conjecture "if n is a poss..."; end n loop */
                    }
                    if (end)
                    {
                        return true;
                    }
                }
            } /* end col loop */
        }
        return end;
    } /* end of this.makeGuess() */

    private guess2() /* desperate now - try the long way */
    {
        //const store: number[][] = initialiseArray(9, 9);
        const storePoss: number[][][] = initialiseArray(9, 9, 9);
        this.log("Beginning an exhaustive backtracking search.");
        for (let i = 0; i < 9; i++) /* fill in the storage arrays */
        {
            for (let j = 0; j < 9; j++)
            {
                //store[i][j] = this.sudokuGrid[i][j];
                for (let k = 0; k < 9; k++)
                {
                    storePoss[i][j][k] = this.possibilities[i][j][k];
                }
            }
        }
        let row = 0;
        let col = 0;
        let n = 1;
        while (row < 9)
        {
            while(col<9)
            {
                while(n<10)
                {
                    if (this.possibilities[row][col][n-1] == 1)
                    {
                        this.sudokuGrid[row][col] = n;
                        n = 1;
                        break;
                    }
                    n++;
                }
                if (n >= 10) //Occurs only if we didn't find anything
                {
                    //Backtrack here
                    if (col === 0)
                    {
                        row--;
                        col = 8;
                    }
                    else
                    {
                        col--;
                    }
                    n = this.sudokuGrid[row][col] + 1;
                    this.sudokuGrid[row][col] = 0;
                    for (let i = 0; i < 9; ++i)
                    {
                        for (let j = 0; j < 9; ++j)
                        {
                            for (let k = 0; k < 9; ++k)
                            {
                                this.possibilities[i][j][k] = storePoss[i][j][k];
                            }
                        }
                    }
                    this.reduce();
                }
                else
                {
                    this.reduce();
                    col++;
                    n = 1;
                }
            }
            row++;
            col = 0;
        }
        // FINISHED
        this.log("Finished.");
    }
}